File: web-source/scripts/sync-notion-subtopics-to-anki.js
const fs = require("fs");
const http = require("http");
const path = require("path");

const projectDir = path.resolve(__dirname, "..", "..");
const exportDir = path.join(projectDir, "project-artifacts", "exports", "anki", "full-export");
const sourceFile = path.join(exportDir, "notion-subtopics-anki-all.tsv");
const resultFile = path.join(exportDir, "notion-subtopics-anki-sync-result.json");
const deckName = "IELTS Topic Collocation::2025-2026 Subtopics";
const chunkSize = 8;

function request(action, params = {}) {
  const payload = JSON.stringify({ action, version: 6, params });

  return new Promise((resolve, reject) => {
    const req = http.request(
      {
        hostname: "127.0.0.1",
        port: 8765,
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Content-Length": Buffer.byteLength(payload)
        },
        timeout: 5000
      },
      (res) => {
        let data = "";
        res.setEncoding("utf8");
        res.on("data", (chunk) => {
          data += chunk;
        });
        res.on("end", () => {
          try {
            const parsed = JSON.parse(data);
            if (parsed.error) {
              reject(new Error(parsed.error));
              return;
            }
            resolve(parsed.result);
          } catch (error) {
            reject(error);
          }
        });
      }
    );

    req.on("timeout", () => {
      req.destroy(new Error("Timed out connecting to AnkiConnect on 127.0.0.1:8765"));
    });
    req.on("error", reject);
    req.write(payload);
    req.end();
  });
}

function parseRows() {
  const lines = fs.readFileSync(sourceFile, "utf8").split(/\r?\n/);
  return lines
    .filter((line) => line && !line.startsWith("#"))
    .map((line) => {
      const cols = line.split("\t");
      if (cols.length !== 11) {
        throw new Error(`Unexpected TSV column count: ${cols.length}`);
      }
      return {
        front: cols[0],
        back: cols[1],
        topic: cols[2],
        topicZh: cols[3],
        subtopic: cols[6],
        tags: cols[9].split(/\s+/).filter(Boolean),
        cardId: cols[10]
      };
    });
}

function buildNote(row) {
  return {
    deckName,
    modelName: "Basic",
    fields: {
      Front: row.front,
      Back: row.back
    },
    tags: row.tags,
    options: {
      allowDuplicate: false,
      duplicateScope: "deck"
    }
  };
}

async function main() {
  const rows = parseRows();

  await request("createDeck", { deck: deckName });
  const notes = rows.map(buildNote);
  const addedNoteIds = [];
  const skippedRows = [];

  for (let start = 0; start < notes.length; start += chunkSize) {
    const noteChunk = notes.slice(start, start + chunkSize);
    const rowChunk = rows.slice(start, start + chunkSize);
    const canAdd = await request("canAddNotes", { notes: noteChunk });
    const addableNotes = noteChunk.filter((_, index) => canAdd[index]);
    skippedRows.push(...rowChunk.filter((_, index) => !canAdd[index]));

    if (addableNotes.length) {
      const chunkNoteIds = await request("addNotes", { notes: addableNotes });
      addedNoteIds.push(...chunkNoteIds);
    }
  }

  const result = {
    syncedAt: new Date().toISOString(),
    deckName,
    sourceFile: path.relative(projectDir, sourceFile),
    totalRows: rows.length,
    added: addedNoteIds.filter(Boolean).length,
    skippedAsDuplicates: skippedRows.length,
    skippedCardIds: skippedRows.map((row) => row.cardId),
    addedNoteIds
  };

  fs.writeFileSync(resultFile, `${JSON.stringify(result, null, 2)}\n`, "utf8");
  console.log(JSON.stringify(result, null, 2));
}

main().catch((error) => {
  const result = {
    syncedAt: new Date().toISOString(),
    deckName,
    sourceFile: path.relative(projectDir, sourceFile),
    error: error.message
  };
  fs.writeFileSync(resultFile, `${JSON.stringify(result, null, 2)}\n`, "utf8");
  console.error(error.message);
  process.exit(1);
});

File: web-source/scripts/sync-notion-subtopics-to-anki.py
import json
import time
import urllib.request
from pathlib import Path

PROJECT_DIR = Path(__file__).resolve().parents[2]
EXPORT_DIR = PROJECT_DIR / "project-artifacts" / "exports" / "anki" / "full-export"
SOURCE_FILE = EXPORT_DIR / "notion-subtopics-anki-all.tsv"
RESULT_FILE = EXPORT_DIR / "notion-subtopics-anki-sync-result.json"
DECK_NAME = "IELTS Topic Collocation::2025-2026 Subtopics"
CHUNK_SIZE = 4


def request(action, params=None, timeout=15):
    payload = json.dumps({"action": action, "version": 6, "params": params or {}}).encode("utf-8")
    req = urllib.request.Request(
        "http://127.0.0.1:8765",
        data=payload,
        headers={"Content-Type": "application/json"},
    )
    with urllib.request.urlopen(req, timeout=timeout) as response:
        data = json.loads(response.read().decode("utf-8"))
    if data.get("error"):
        raise RuntimeError(data["error"])
    return data.get("result")


def parse_rows():
    rows = []
    for line in SOURCE_FILE.read_text(encoding="utf-8").splitlines():
        if not line or line.startswith("#"):
            continue
        cols = line.split("\t")
        if len(cols) != 11:
            raise RuntimeError(f"Unexpected TSV column count: {len(cols)}")
        rows.append(
            {
                "front": cols[0],
                "back": cols[1],
                "tags": [tag for tag in cols[9].split() if tag],
                "cardId": cols[10],
            }
        )
    return rows


def build_note(row):
    return {
        "deckName": DECK_NAME,
        "modelName": "Basic",
        "fields": {
            "Front": row["front"],
            "Back": row["back"],
        },
        "tags": row["tags"],
        "options": {
            "allowDuplicate": False,
            "duplicateScope": "deck",
        },
    }


def main():
    rows = parse_rows()
    notes = [build_note(row) for row in rows]
    added_note_ids = []
    skipped_rows = []

    request("createDeck", {"deck": DECK_NAME})

    for start in range(0, len(notes), CHUNK_SIZE):
        note_chunk = notes[start : start + CHUNK_SIZE]
        row_chunk = rows[start : start + CHUNK_SIZE]
        can_add = request("canAddNotes", {"notes": note_chunk})
        addable_notes = [note for note, ok in zip(note_chunk, can_add) if ok]
        skipped_rows.extend(row for row, ok in zip(row_chunk, can_add) if not ok)
        if addable_notes:
            added_note_ids.extend(request("addNotes", {"notes": addable_notes}))
        time.sleep(0.2)

    result = {
        "syncedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "deckName": DECK_NAME,
        "sourceFile": str(SOURCE_FILE.relative_to(PROJECT_DIR)),
        "totalRows": len(rows),
        "added": len([note_id for note_id in added_note_ids if note_id]),
        "skippedAsDuplicates": len(skipped_rows),
        "skippedCardIds": [row["cardId"] for row in skipped_rows],
        "addedNoteIds": added_note_ids,
    }
    RESULT_FILE.write_text(json.dumps(result, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        result = {
            "syncedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "deckName": DECK_NAME,
            "sourceFile": str(SOURCE_FILE.relative_to(PROJECT_DIR)),
            "error": str(error),
        }
        RESULT_FILE.write_text(json.dumps(result, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
        raise
